cmake/add_cpp_test.cmake

```diff
@@ -1,38 +1,69 @@
 # Helper so we don't have to repeat ourselves so much
-# Usage: add_cpp_test(testname [COMPILE_ONLY] [SOURCES a.cpp b.cpp ...] [LABELS acceptance per_implementation ...])
-# SOURCES defaults to testname.cpp if not specified.
+# Usage: add_cpp_test(<testname>
+#                     [COMPILE_ONLY]
+#                     [DEBUG_ONLY]
+#                     [WILL_FAIL]
+#                     [LIBRARY]
+#                     [SOURCES a.cpp b.cpp ...]
+#                     [LABELS acceptance per_implementation ...]
+#                     [DEPENDENCY_OF ...])
+# SOURCES defaults to <testname>.cpp if not specified.
 function(add_cpp_test TEST_NAME)
   # Parse arguments
-  cmake_parse_arguments(PARSE_ARGV 1 ARGS "COMPILE_ONLY;LIBRARY;WILL_FAIL" "" "SOURCES;LABELS;DEPENDENCY_OF")
-  if (NOT ARGS_SOURCES)
+  cmake_parse_arguments(PARSE_ARGV 1 ARGS "COMPILE_ONLY;LIBRARY;WILL_FAIL;DEBUG_ONLY" "" "SOURCES;LABELS;DEPENDENCY_OF")
+
+  # Is generator multi-config?
+  get_cmake_property(is_multi_config GENERATOR_IS_MULTI_CONFIG)
+
+  # Return if debug only test
+  if(NOT is_multi_config AND ARGS_DEBUG_ONLY AND NOT CMAKE_BUILD_TYPE STREQUAL "Debug")
+    return()
+  endif()
+
+  # Set configurations
+  set(config_args)
+  if(is_multi_config AND ARGS_DEBUG_ONLY)
+    set(config_args CONFIGURATIONS Debug)
+  endif()
+
+  if(NOT ARGS_SOURCES)
     list(APPEND ARGS_SOURCES ${TEST_NAME}.cpp)
   endif()
-  if (ARGS_COMPILE_ONLY)
-    list(APPEND ${ARGS_LABELS} compile_only)
+
+  # Wrap sources in a generator expression and provide a dummy source
```

Comment:
Oh, now I understand it. The distance between this statement and the setting of "is_multi_config" meant I didn't realize these were related. Perhaps if this if() statement and the previous one were both in the same place, this would be easier to grok?

A comment explaining the purpose is still probably warranted.

Reply:
I was hoping the condition coupled with the comment was enough. This block basically makes sure that in configurations other than Debug, the only source this target will compile is the empty cpp file. There is no way to disable targets cleanly based on configuration, so just using a dummy source to compile instead of the actual source(s) was the best option.